satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,387 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gracefulexit
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/rpc/rpcstatus"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/metainfo"
+	"storj.io/storj/satellite/orders"
+	"storj.io/storj/satellite/overlay"
+	"storj.io/storj/uplink/eestream"
+)
+
+// drpcEndpoint wraps streaming methods so that they can be used with drpc
+type drpcEndpoint struct{ *Endpoint }
+
+// processStream is the minimum interface required to process requests.
+type processStream interface {
+	Context() context.Context
+	Send(*pb.SatelliteMessage) error
+	Recv() (*pb.StorageNodeMessage, error)
+}
+
+// Endpoint for handling the transfer of pieces for Graceful Exit.
+type Endpoint struct {
+	log       *zap.Logger
+	interval  time.Duration
+	db        DB
+	overlaydb overlay.DB
+	overlay   *overlay.Service
+	metainfo  *metainfo.Service
+	orders    *orders.Service
+	config    Config
+}
+
+type pendingTransfer struct {
+	path             []byte
+	pieceSize        int64
+	satelliteMessage *pb.SatelliteMessage
+}
+
+// DRPC returns a DRPC form of the endpoint.
+func (endpoint *Endpoint) DRPC() pb.DRPCSatelliteGracefulExitServer {
+	return &drpcEndpoint{Endpoint: endpoint}
+}
+
+// NewEndpoint creates a new graceful exit endpoint.
+func NewEndpoint(log *zap.Logger, db DB, overlaydb overlay.DB, overlay *overlay.Service, metainfo *metainfo.Service, orders *orders.Service, config Config) *Endpoint {
+	return &Endpoint{
+		log:       log,
+		interval:  time.Millisecond * 100,
```

Comment:
This should probably be a const or have a TODO

Reply:
changed to const